conf/schemas.go

```diff
@@ -58,50 +91,114 @@ func ReadSchemas(file string) (Schemas, error) {
 		}
 
 		if sec.ValueOf("pattern") == "" {
-			return nil, fmt.Errorf("[%s]: empty pattern", schema.Name)
+			return Schemas{}, fmt.Errorf("[%s]: empty pattern", schema.Name)
 		}
 		schema.Pattern, err = regexp.Compile(sec.ValueOf("pattern"))
 		if err != nil {
-			return nil, fmt.Errorf("[%s]: failed to parse pattern %q: %s", schema.Name, sec.ValueOf("pattern"), err.Error())
+			return Schemas{}, fmt.Errorf("[%s]: failed to parse pattern %q: %s", schema.Name, sec.ValueOf("pattern"), err.Error())
 		}
 
 		schema.Retentions, err = ParseRetentions(sec.ValueOf("retentions"))
 		if err != nil {
-			return nil, fmt.Errorf("[%s]: failed to parse retentions %q: %s", schema.Name, sec.ValueOf("retentions"), err.Error())
+			return Schemas{}, fmt.Errorf("[%s]: failed to parse retentions %q: %s", schema.Name, sec.ValueOf("retentions"), err.Error())
 		}
 
 		p := int64(0)
 		if sec.ValueOf("priority") != "" {
 			p, err = strconv.ParseInt(sec.ValueOf("priority"), 10, 0)
 			if err != nil {
-				return nil, fmt.Errorf("Failed to parse priority %q for [%s]: %s", sec.ValueOf("priority"), schema.Name, err)
+				return Schemas{}, fmt.Errorf("Failed to parse priority %q for [%s]: %s", sec.ValueOf("priority"), schema.Name, err)
 			}
 		}
 		schema.Priority = int64(p)<<32 - int64(i) // to sort records with same priority by position in file
 
 		schemas = append(schemas, schema)
 	}
 
-	sort.Sort(schemas)
-	return schemas, nil
+	sort.Sort(SchemaSlice(schemas))
+
+	return NewSchemas(schemas), nil
 }
 
 // Match returns the correct schema setting for the given metric
 // it can always find a valid setting, because there's a default catch all
 // also returns the index of the setting, to efficiently reference it
-func (s Schemas) Match(metric string) (uint16, Schema) {
-	for i, schema := range s {
+func (s Schemas) Match(metric string, interval int) (uint16, Schema) {
+	i := 0
+	for i < len(s.index) {
+		schema := s.index[i]
 		if schema.Pattern.MatchString(metric) {
-			return uint16(i), schema
+			// no interval passed,use the raw retentions.
+			// This is primarily used by the carbon input plugin.
+			if interval == 0 {
+				return uint16(i), *schema
+			}
+			// search through the retentions to find the first one where
+			// the metric interval is < SecondsPerPoint of the retention.
+			// The schema is then the previous retention.
+			for j, ret := range schema.Retentions {
+				if interval < ret.SecondsPerPoint {
+					// if there are no retentions with SecondsPerPoint >= interval
+					// then we need to use the first retention. Otherwise, the retention
+					// we want to use is the previous one.
```

Comment:
> if a < b is true, then b >= a is false.

no, these are almost equivalent. e.g. if a is 1 and b is 2 then both are true.

> ie these are the same thing

no they are not:
```
	interval := 10
	SecondsPerPoint := 20
	fmt.Println(interval < SecondsPerPoint)
	fmt.Println(!(SecondsPerPoint >= interval))
```

this prints:
```
true
false
```